src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,1010 @@ mod v1 {
+    use std::time::Instant;
+    use crc::crc32::{self, Digest, Hasher32};
+    use protobuf::{Message, RepeatedField};
+    use kvproto::metapb::Region;
+    use kvproto::raft_serverpb::{SnapshotCFFile, SnapshotMeta, RaftSnapshotData};
+    use rocksdb::{EnvOptions, SstFileWriter, IngestExternalFileOptions};
+    use storage::{CfName, CF_LOCK};
+    use util::{HandyRwLock, rocksdb, duration_to_sec};
+    use util::file::{get_file_size, file_exists, delete_file_if_exist};
+    use raftstore::Result as RaftStoreResult;
+
+    use super::super::engine::{Snapshot as DbSnapshot, Iterable};
+    use super::super::keys::{enc_start_key, enc_end_key};
+    use super::super::metrics::{SNAPSHOT_CF_KV_COUNT, SNAPSHOT_CF_SIZE,
+                                SNAPSHOT_BUILD_TIME_HISTOGRAM};
+    use super::{SNAPSHOT_CFS, SNAP_GEN_PREFIX, SNAP_REV_PREFIX, TMP_FILE_SUFFIX, SST_FILE_SUFFIX,
+                Result, SnapKey, Snapshot, BuildContext, ApplyOptions, check_abort};
+    use super::v1::{build_plain_cf_file, apply_plain_cf_file};
+
+    pub const SNAPSHOT_VERSION: u64 = 2;
+    const META_FILE_SUFFIX: &'static str = ".meta";
+    const DIGEST_BUFFER_SIZE: usize = 10240;
+
+
+    #[inline]
+    fn plain_file_used(cf: &str) -> bool {
+        cf == CF_LOCK
+    }
+
+    fn calc_crc32(p: &PathBuf) -> io::Result<u32> {
+        let mut digest = Digest::new(crc32::IEEE);
+        let mut f = try!(OpenOptions::new().read(true).open(&p));
+        let mut buf = vec![0; DIGEST_BUFFER_SIZE];
+        loop {
+            match f.read(&mut buf[..]) {
+                Ok(0) => {
+                    return Ok(digest.sum32());
+                }
+                Ok(n) => {
+                    digest.write(&buf[..n]);
+                }
+                Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+                Err(err) => return Err(err),
+            }
+        }
+    }
+
+    fn get_snapshot_meta(cf_files: &[CfFile]) -> RaftStoreResult<SnapshotMeta> {
+        let mut meta = Vec::with_capacity(cf_files.len());
+        for cf_file in cf_files {
+            if SNAPSHOT_CFS.iter().find(|&cf| cf_file.cf == *cf).is_none() {
+                return Err(box_err!("failed to encode invalid snapshot cf {}", cf_file.cf));
+            }
+
+            let mut cf_file_meta = SnapshotCFFile::new();
+            cf_file_meta.set_cf(cf_file.cf.to_owned());
+            cf_file_meta.set_size(cf_file.size);
+            cf_file_meta.set_checksum(cf_file.checksum);
+            meta.push(cf_file_meta);
+        }
+        let mut snapshot_meta = SnapshotMeta::new();
+        snapshot_meta.set_cf_files(RepeatedField::from_vec(meta));
+        Ok(snapshot_meta)
+    }
+
+    fn check_snapshot_meta(snapshot_meta: &SnapshotMeta) -> RaftStoreResult<Vec<CfName>> {
+        let mut res = Vec::with_capacity(SNAPSHOT_CFS.len());
+        for cf_file_meta in snapshot_meta.get_cf_files() {
+            let found = SNAPSHOT_CFS.iter().find(|&s| cf_file_meta.get_cf() == *s);
```

Comment:
It is to use the global references in `SNAPSHOT_CFS`, although the `snapshot_meta` is cached in `Snap`, it's fine to use cf reference of it.